hazelcast/src/main/java/com/hazelcast/nio/tcp/WriteHandler.java

```diff
@@ -172,7 +172,7 @@ private void unschedule() {
         // So the outputBuffer is empty, so we are going to unschedule ourselves.
         scheduled.set(false);
 
-        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty()) {
+        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty() && currentPacket == null) {
```

Comment:
It is difficult one. The problem is that if there would not be some kind of regular packet that released the stuck packet, you would get a testing failure. But because e.g. a heart beat and other potential operations (responses etc) the error gets hidden.

A general purpose performance test where we test with small packets and many concurrently won't expose the problem. Only with large packets and little concurrent calls, this issue would have showed up. But.. probably only a performance degradation issue.. not from a correctness point of view (due to e..g heart beat).

My crucial mistake here is that I did refactoring without sufficient tests being available for the IO system. The IO System relies on implicit tests; so it gets tested because other components get tested that rely on the io system (e.g. the IMap).

So in this case we should have provided a lot of tests before attempting to refactoring.  But to make this happen, we need to make sure that we get better testable code. And this by its very nature, requires changes in the actual code to make it happen.